### PR TITLE
Iox #996 add expect to expected and optional

### PR DESCRIPTION
**Pull request by elfenpiff**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CHANGELOG.md

## Notes for Reviewer
This is the first pull request more will follow. The goal for this PR was to provide an interface class with tests which enables us to reduce a lot of code duplication (implementation and testing) in methods like `and_then`, `or_else`, `value_or`, `expect` for `optional`, `expected`, `shared_ptr`, `unique_ptr` etc...

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
    - [x] Each unit test case has a unique UUID
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes **TBD**

### CODE REVIEW

**Comment by elBoberido** [comment]
Would you mind to restructure this PR? I know it would cost you some time but it is quite large and I wouldn't be able to review this in one go anyway. Splitting this up might in the end be faster than trying to get this in all at once. You could maybe try to create stacked PRs with branches on eclipse-iceoryx. My proposal would be
1. pick one function, e.g. `AndThen`
2. create a the necessary files with the implementation
3. create the test by e.g. `test_functional_interface_common.cpp` and `test_functional_interface_and_then.cpp`
4. create PR
5. repeat with 1

Currently there are multiple typed tests in there and it is quite hard to follow, especially since the PR is quite large. With the proposal above you would be able to merge some parts while the other would still be in review. If you want, I can give you some suggestion how this workflow can be made quite effective.

**Comment by elfenpiff** [comment]
@elBoberido I split up the tests into multiple files so that every functional interface has its own test file and an additional test file was added where I concatenate multiple calls in one test.
I hope it eases your pain when the test files are small and grouped.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/functional_interface.hpp [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+struct HasGetErrorMethod<T, cxx::void_t<decltype(std::declval<T>().get_error())>> : std::true_type
+{
+};
+
+template <typename T>
+struct Expect
+{
+    /// @brief Expects that the object is valid, otherwise the method prints the
+    ///        provided message and induces a fatal error
+    /// @param[in] msg Message which will be print when the object is invalid
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/include/iceoryx_hoofs/cxx/functional_interface.hpp:54]:
```suggestion
    /// @param[in] msg Message which will be printed when the object is invalid
```
Please correct this also in all comments in the `ExpectWithValue` struct.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/functional_interface.hpp [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+    ValueType value_or(const ValueType& alternative) const noexcept;
+};
+
+template <typename T, typename ValueType>
+struct AndThenWithValue
+{
+    using and_then_callback_t = cxx::function_ref<void(ValueType&)>;
+    using const_and_then_callback_t = cxx::function_ref<void(const ValueType&)>;
+
+    /// @brief Calls the provided callable when the object is valid and provide the underlying
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/include/iceoryx_hoofs/cxx/functional_interface.hpp:103]:
```suggestion
    /// @brief Calls the provided callable when the object is valid and provides the underlying
```
Please correct this also in the following comments.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/functional_interface.hpp [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+    void expect(const char* const msg) const noexcept;
+};
+
+template <typename T, typename ValueType>
+struct ExpectWithValue
+{
+    /// @brief Expects that the object is valid and returns the contained value, otherwise
+    //         the method prints the provided message and induces a fatal error
+    /// @param[in] msg Message which will be print when the object is invalid
+    /// @return the contained value type
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/include/iceoryx_hoofs/cxx/functional_interface.hpp:64]:
```suggestion
    /// @return a reference to the contained value
```

**Review by FerdinandSpitzschnueffler — COMMENTED** [review]
Just started to get an overview.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/functional_interface.inl [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+namespace iox
+{
+namespace cxx
+{
+namespace internal
+{
+template <typename T>
+inline void Expect<T>::expect(const char* const msg) const noexcept
+{
+    if (!(*static_cast<const T*>(this)))
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/functional_interface.inl:29]: What do you think of renaming `T` to `Derived`? This would make it clearer what happens here (at least for me).

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/functional_interface.inl [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+}
+
+template <typename T, typename ValueType>
+inline const ValueType&& ExpectWithValue<T, ValueType>::expect(const char* const msg) const&& noexcept
+{
+    return const_cast<const ValueType&&>(std::move(const_cast<ExpectWithValue<T, ValueType>*>(this)->expect(msg)));
+}
+
+template <typename T, typename ValueType>
+inline ValueType ValueOr<T, ValueType>::value_or(const ValueType& value) const noexcept
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/functional_interface.inl:69]: No need to change, but I think readability would improve when you rename `value` to `alternative` like you did in the header file.

**Review by FerdinandSpitzschnueffler — COMMENTED** [review]
Didn't reviewed the tests yet.

File: iceoryx_hoofs/test/moduletests/test_cxx_functional_interface_and_then.cpp [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+        });
+        EXPECT_TRUE(wasCallbackCalled);
+    }
+};
+
+TYPED_TEST(FunctionalInterface_test, AndThenIsCalledCorrectlyWhenValid_LValueCase)
+{
+    using SutType = typename TestFixture::TestFactoryType::Type;
+    AndThenIsCalledCorrectlyWhenValid<iox::cxx::internal::HasValueMethod<SutType>::value>::template performTest<
+        typename TestFixture::TestFactoryType>([](auto& sut, auto callback) { sut.and_then(callback); });
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/test/moduletests/test_cxx_functional_interface_and_then.cpp:59]: Just want to make sure that I get it right:
1. `HasValueMethod` defined in `functional_interface.hpp` evaluates to true/false depending on whether `SutType` has a `value()` method or not. The specialization of `AndThenIsCalledCorrectlyWhenValid` is selected accordingly.
2. `performTest` is called with `[](auto& sut, auto callback) { sut.and_then(callback); }` as argument
3. in `performTest` a valid object is created (=`sut`)
4. the callback argument `andThenCall` is executed like that:
```
sut.and_then([&](auto& arg) {
    wasCallbackCalled = true;
    EXPECT_EQ(arg, TestFactory::usedTestValue);
});
```
with `sut` from 3. and `AndThenIsCalledCorrectlyWhenValid<true>`.

Please let me know if my understanding is correct.

File: iceoryx_hoofs/test/moduletests/test_cxx_functional_interface_expect.cpp [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+{
+    ExpectDoesCallTerminateWhenObjectIsInvalid<typename TestFixture::TestFactoryType>(
+        [](auto& sut) { sut.expect("the chocolate rations will be increased soon"); });
+}
+
+TYPED_TEST(FunctionalInterface_test, ExpectDoesCallTerminateWhenObjectIsInvalid_constLValueCase)
+{
+    using SutType = typename TestFixture::TestFactoryType::Type;
+    ExpectDoesCallTerminateWhenObjectIsInvalid<typename TestFixture::TestFactoryType>(
+        [](auto& sut) { const_cast<const SutType&>(sut).expect("hypnotoad ate the spagetti monster"); });
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/test/moduletests/test_cxx_functional_interface_expect.cpp:93]:
```suggestion
        [](auto& sut) { const_cast<const SutType&>(sut).expect("hypnotoad ate the spaghetti monster"); });
```

File: iceoryx_hoofs/test/moduletests/test_cxx_functional_interface_expect.cpp [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+}
+
+TYPED_TEST(FunctionalInterface_test, ExpectDoesCallTerminateWhenObjectIsInvalid_RValueCase)
+{
+    ExpectDoesCallTerminateWhenObjectIsInvalid<typename TestFixture::TestFactoryType>(
+        [](auto& sut) { std::move(sut).expect("the spagetti monster ate hypnotoad"); });
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/test/moduletests/test_cxx_functional_interface_expect.cpp:99]:
```suggestion
        [](auto& sut) { std::move(sut).expect("the spaghetti monster ate hypnotoad"); });
```

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/functional_interface.hpp [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+template <typename Derived, typename ValueType>
+struct AndThenWithValue
+{
+    using and_then_callback_t = cxx::function_ref<void(ValueType&)>;
+    using const_and_then_callback_t = cxx::function_ref<void(const ValueType&)>;
+
+    /// @brief Calls the provided callable when the object is valid and provides the underlying
+    ///        value reference as argument to the callable. If the object is not valid, nothing
+    ///        happens.
+    /// @param[in] callable Callable to be called when valid
```
Comment (elBoberido) [inline iceoryx_hoofs/include/iceoryx_hoofs/cxx/functional_interface.hpp:106]: Not quite sure how this will be rendered and I think we do not follow a common style here but is looks a bit weird to repeat the `callable` as `Callable`.

Don't care that much which style to use but it should be consistent. Do you know it clang tidy could check this for us or do we (finger on my nose) one day need to check the whole documentation for this?

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/requires.hpp [hunk 10]
```diff
@@ -0,0 +1,45 @@
+// Copyright (c) 2022 by Apex.AI Inc. All rights reserved.
```
Comment (elBoberido) [inline iceoryx_hoofs/include/iceoryx_hoofs/cxx/requires.hpp:1]: I'm pretty sure you have to add the Bosch copyright here

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/requires.hpp [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+namespace internal
+{
+void Require(
+    const bool condition, const char* file, const int line, const char* function, const char* conditionString) noexcept;
+} // namespace internal
+
+// implementing C++ Core Guideline, I.6. Prefer Expects
+// see:
+// https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Ri-expects
+#define Expects(condition) internal::Require(condition, __FILE__, __LINE__, __PRETTY_FUNCTION__, #condition)
```
Comment (elBoberido) [inline iceoryx_hoofs/include/iceoryx_hoofs/cxx/requires.hpp:34]: It is really unfortunate that macros are not namespaced. While this macro will produce an error when used without the `iox` namespace, it can still cause issues when someone else also has an `Expects` macro. It is really a pity that the core guidelines do not recommend to prefix this macro in order to prevent collisions.

It is nothing for the PR, but do you think it would be worth to rename this to `IOX_EXPECTS`?

File: iceoryx_hoofs/source/cxx/requires.cpp [hunk 12]
```diff
@@ -0,0 +1,40 @@
+// Copyright (c) 2022 by Apex.AI Inc. All rights reserved.
```
Comment (elBoberido) [inline iceoryx_hoofs/source/cxx/requires.cpp:1]: Please also add the Bosch copyright since this was copied from the helplets

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/functional_interface.inl [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+        Ensures(false);
+    }
+
+    return derivedThis->value();
+}
+
+template <typename Derived, typename ValueType>
+inline const ValueType& ExpectWithValue<Derived, ValueType>::expect(const char* const msg) const& noexcept
+{
+    return const_cast<const ValueType&>(const_cast<ExpectWithValue<Derived, ValueType>*>(this)->expect(msg));
```
Comment (elBoberido) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/functional_interface.inl:53]: What do you think about having a `using Self = ExpectWithValue<Derived, ValueType>` alias in the class definition and use it here?

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/functional_interface.inl [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+inline ValueType&& ExpectWithValue<Derived, ValueType>::expect(const char* const msg) && noexcept
+{
+    return std::move(this->expect(msg));
+}
+
+template <typename Derived, typename ValueType>
+inline const ValueType&& ExpectWithValue<Derived, ValueType>::expect(const char* const msg) const&& noexcept
+{
+    return const_cast<const ValueType&&>(
+        std::move(const_cast<ExpectWithValue<Derived, ValueType>*>(this)->expect(msg)));
```
Comment (elBoberido) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/functional_interface.inl:66]: Could also be used here

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/functional_interface.inl [hunk 15]
```diff
@@ -0,0 +1,228 @@
...
+{
+    return const_cast<const Derived&>(const_cast<OrElse<Derived>*>(this)->or_else(callable));
+}
+
+template <typename Derived>
+inline const Derived&& OrElse<Derived>::or_else(const or_else_callback_t& callable) const&& noexcept
+{
+    return std::move(this->or_else(callable));
+}
+
```
Comment (elBoberido) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/functional_interface.inl:223]: Could you please add
```cpp
// BEGIN and_then
...
// END and_then
```
comments for the different interfaces? This helps to distinguish where the implementation of a new interface begins and also enables code folding on many IDEs.

**Review by elBoberido — COMMENTED** [review]
Tests not yet reviewed. Implementation looks good overall

File: iceoryx_hoofs/test/moduletests/test_cxx_functional_interface_or_else.cpp [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+TYPED_TEST(FunctionalInterface_test, OrElseIsCalledCorrectlyWhenInvalid_ConstRValueCase)
+{
+    IOX_TEST(OrElseIsCalledCorrectlyWhenInvalid, std::move(const_cast<const SutType&>(sut)));
+}
+
+template <bool HasError>
+struct OrElseIsNotCalledWhenValid;
+
+template <>
+struct OrElseIsNotCalledWhenValid<false>
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/test/moduletests/test_cxx_functional_interface_or_else.cpp:87]:
```suggestion
struct OrElseIsNotCalledWhenValid<TYPE_HAS_NO_GET_ERROR_METHOD>
```

File: CHANGELOG.md [hunk 17]
```diff
@@ -59,6 +59,7 @@
 - Remove `ApplicationPort` and `ApplicationPortData` classes [\#415](https://github.com/eclipse-iceoryx/iceoryx/issues/415)
 - Remove creation pattern from `MemoryMap` and replace it with `MemoryMapBuilder` [\#1036](https://github.com/eclipse-iceoryx/iceoryx/issues/1036)
 - Fix error handling of `TypedUniqueId` and refactor it to `UniquePortId` [\#861](https://github.com/eclipse-iceoryx/iceoryx/issues/861)
+- `cxx::Expects` and `cxx::Ensures` are moved into the header `requires.hpp` [\#996](https://github.com/eclipse-iceoryx/iceoryx/issues/996)
```
[the remark this answers is not in the document]
Comment (elfenpiff) [reply CHANGELOG.md:62]: I rewrote the issue so that it states in more detail what, how and why we are doing here.

@elBoberido but I would leave this line as it is in the changelog? What is your concern here? It states exactly what has changed in the public API.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/functional_interface.hpp [hunk 18]
```diff
@@ -0,0 +1,279 @@
...
+
+template <typename Derived, typename ValueType>
+struct ValueOr
+{
+    /// @brief When the object contains a value a copy will be returned otherwise a
+    ///        a copy of alternative will be returned
+    /// @param[in]  alternative the return value which will be used when the object does
+    ///             not contain a value
+    /// @return A copy of the contained value if possible otherwise a copy of alternative
+    ValueType value_or(const ValueType& alternative) const noexcept;
```
[the remark this answers is not in the document]
Comment (elfenpiff) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/functional_interface.hpp:94]: You are right, I will change this

File: CHANGELOG.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (elBoberido) [reply CHANGELOG.md:62]: It is just that this PR also adds the functional interface and that it should be part of the chagnelog

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/functional_interface.hpp (hunk 9 above)
Comment (elfenpiff) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/functional_interface.hpp:118]: I rephrased it into less weird

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/requires.hpp (hunk 11 above)
Comment (elfenpiff) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/requires.hpp:34]: Let's wait how the `Ensures`, `Expects` or `Requires` discussion ends and then we may have only an `IOX_REQUIRES` or something completely different.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/functional_interface.inl (hunk 13 above)
Comment (elfenpiff) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/functional_interface.inl:53]: I added the Self alias in the method scope here and in some other places.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/functional_interface.hpp [hunk 19]
```diff
@@ -0,0 +1,279 @@
...
+{
+};
+
+template <typename Derived>
+struct Expect
+{
+    /// @brief Expects that the object is valid, otherwise the method prints the
+    ///        provided message and induces a fatal error
+    /// @param[in] msg Message which will be printed when the object is invalid
+    void expect(const char* const msg) const noexcept;
```
[the remark this answers is not in the document]
Comment (elfenpiff) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/functional_interface.hpp:56]: I'll keep your proposal open to think a little bit about it.

File: iceoryx_hoofs/test/moduletests/test_cxx_functional_interface_and_then.cpp [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+#define IOX_TEST(TestName, variationPoint)                                                                             \
+    using SutType = typename TestFixture::TestFactoryType::Type;                                                       \
+    TestName<iox::cxx::internal::HasValueMethod<SutType>::value>::template performTest<                                \
+        typename TestFixture::TestFactoryType>([](auto& sut, auto callback) { variationPoint.and_then(callback); })
+
+constexpr bool TYPE_HAS_VALUE_METHOD = true;
+constexpr bool TYPE_HAS_NO_VALUE_METHOD = false;
+
+template <bool HasValue>
+struct AndThenIsCalledCorrectlyWhenValid;
```
Comment (elBoberido) [inline iceoryx_hoofs/test/moduletests/test_cxx_functional_interface_and_then.cpp:32]: Would it be possible to add a static_assert with a reasonable error massage that the type has not the appropriate methods?
I think it should be possible like here
```cpp
template <bool HasValue, typename SutType>
struct AndThenIsCalledCorrectlyWhenValid {
    static_assert(always_false_v(SutType) && "The type does not implement a `value` method!");
};
```

File: iceoryx_hoofs/test/moduletests/test_cxx_functional_interface_and_then.cpp [hunk 21]
```diff
@@ -0,0 +1,137 @@
...
+#include "test_cxx_functional_interface_types.hpp"
+
+namespace
+{
+using namespace test_cxx_functional_interface;
+using namespace ::testing;
+
+#define IOX_TEST(TestName, variationPoint)                                                                             \
+    using SutType = typename TestFixture::TestFactoryType::Type;                                                       \
+    TestName<iox::cxx::internal::HasValueMethod<SutType>::value>::template performTest<                                \
```
Comment (elBoberido) [inline iceoryx_hoofs/test/moduletests/test_cxx_functional_interface_and_then.cpp:25]: To make this a bit more readable, What do you think about the following?
```suggestion
    constexpr auto HAS_VALUE_METHOD = iox::cxx::internal::HasValueMethod<SutType>::value; \
    TestName<HAS_VALUE_METHOD>::template performTest<                                                       \
```
Maybe with a better name for the constexpr?

It would reduce the cognitive load when parsing in that line

File: iceoryx_hoofs/test/moduletests/test_cxx_functional_interface_types.hpp [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+///
+///        * static error_t usedErrorValue;
+///            Error value which was used while creating an invalid object
+///
+///        * static error_t anotherErrorValue
+///            Another error value which can be compared to usedErrorValue and is not equal to it
+
+using FunctionalInterfaceImplementations = testing::Types<GenericValueErrorFactory, GenericPlainFactory>;
+TYPED_TEST_SUITE(FunctionalInterface_test, FunctionalInterfaceImplementations);
+
```
Comment (elBoberido) [inline iceoryx_hoofs/test/moduletests/test_cxx_functional_interface_types.hpp:95]: What I'm missing is a check that the expected functional interface is implemented, e.g. that `cxx::optional` has `and_then` with value and `or_else` without value and not the other way around.
In the end, this will catch ones eye when the `optional` is used but that is also the case for many other things we test and since there is quite some template magic involved I think we should ensure that the expected interfaces are implemented. I think it might even be possible to test whether the wrong interfaces are not implemented. This should be doable by adding some `constexpr bool EXPECT_AND_THEN_WITH_VALUE {true}` or similar to the factory class.

**Review by elBoberido — COMMENTED** [review]
Reviewed the generic test classes and the `and_then` tests

File: iceoryx_hoofs/test/moduletests/test_cxx_functional_interface_types.hpp [hunk 23]
```diff
@@ -0,0 +1,98 @@
...
+///            Type alias of the error type
+///
+///        * static error_t usedErrorValue;
+///            Error value which was used while creating an invalid object
+///
+///        * static error_t anotherErrorValue
+///            Another error value which can be compared to usedErrorValue and is not equal to it
+
+using FunctionalInterfaceImplementations = testing::Types<GenericValueErrorFactory, GenericPlainFactory>;
+TYPED_TEST_SUITE(FunctionalInterface_test, FunctionalInterfaceImplementations);
```
[the remark this answers is not in the document]
Comment (elfenpiff) [reply iceoryx_hoofs/test/moduletests/test_cxx_functional_interface_types.hpp:94]: I think the test would be more complex then and the amount of  code a user has to write would also not decrease. I would dismiss it for now but when I integrate optional and expected then I will try to keep this in mind and I see some nifty approach to simplify it even more.

File: iceoryx_hoofs/test/moduletests/test_cxx_functional_interface_and_then.cpp (hunk 20 above)
Comment (elfenpiff) [reply iceoryx_hoofs/test/moduletests/test_cxx_functional_interface_and_then.cpp:32]: I think this does not make sense or maybe I misunderstood you.
It is completely valid for a type to not implement a value method. The test if the `and_then` and `or_else` signature is correct has to be done then in the types tests for instance in the optional tests.

Comment (elfenpiff) [reply iceoryx_hoofs/test/moduletests/test_cxx_functional_interface_and_then.cpp:41]: I am confused?! I think this static assert would never been called since it cannot happen that a sut without value method is called in such a struct. I would use the same means in the static assert as I would use in the specialization of the struct. `HasValueMethod<SutType>::value`

**Review by elBoberido — COMMENTED** [review]
Looks good so far. I still think that `IOX_TEST` is too generic. There are multiple ways this can break in the future, e.g. by having a `IOX_TEST` macro which wraps the gTest `TEST` macro which does some repetitive stuff, like controlling the log level.

Furthermore, it is quite easy to forget the `#undef IOX_TEST` at the end of the file. Oh btw, you forgot a `#undef IOX_TEST` in one of the tests.

It is not a hard request to change this, I just want to remind you of the pain you had with some macros on Windows.

**Review by elBoberido — APPROVED** [review]

**Review by FerdinandSpitzschnueffler — APPROVED** [review]

**Review by FerdinandSpitzschnueffler — CHANGES_REQUESTED** [review]
Please add unique test ids.

**Review by FerdinandSpitzschnueffler — APPROVED** [review]

**Review by elBoberido — APPROVED** [review]
